internal/provider/resource_access_generic.go

```diff
@@ -140,27 +155,290 @@ func (a *genericJAASAccessResource) Configure(ctx context.Context, req resource.
+func (resource *genericJAASAccessResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Check first if the client is configured
+	if resource.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, resource.resourceLogName, "read")
+		return
+	}
+	// Read Terraform configuration from the request into the plan
+	state, targetTag := resource.info(ctx, req.State, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
+	// Create a tuple that defines what relations we are interested in
+	readTuple := juju.JaasTuple{
+		Target:   targetTag.String(),
+		Relation: state.Access.ValueString(),
+	}
+	tuples, err := resource.client.Jaas.ReadRelations(ctx, &readTuple)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read access rules, got error: %s", err))
+		return
+	}
+	// Transform the tuples into an access model
+	newModel := tuplesToModel(ctx, tuples, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Users = newModel.Users
+	state.Groups = newModel.Groups
+	state.ServiceAccounts = newModel.ServiceAccounts
+	resp.Diagnostics.Append(resource.targetResource.Save(ctx, &resp.State, state, targetTag)...)
 }
 
 // Update defines how tuples for access control will be updated.
-func (a *genericJAASAccessResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+func (resource *genericJAASAccessResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if resource.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, resource.resourceLogName, "update")
+		return
+	}
+
+	// Note: We only need to read the targetID from either the plan or the state.
+	// If it changed, the resource should be replaced rather than updated.
+
+	// Read Terraform configuration from the state
+	state, targetTag := resource.info(ctx, req.State, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read Terraform configuration from the plan
+	plan, _ := resource.info(ctx, req.Plan, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get a diff of the plan vs. state to know what relations to add/remove
+	modelAdd, modelRemove := diffModels(plan, state, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create a list of tuples to add and tuples to remove
+	addTuples := modelToTuples(ctx, targetTag, modelAdd, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	removeTuples := modelToTuples(ctx, targetTag, modelRemove, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Add new relations
+	if len(addTuples) > 0 {
+		err := resource.client.Jaas.AddRelations(addTuples)
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to add access rules, got error: %s", err))
+			return
```

Comment:
question: is it a given that failing to AddRelations indicates that DeleteRelations will also fail? can you continue rather than return and attempt the Delete?

Reply:
We could try to continue but then if DeleteRelations suceeds, the TF state will be incorrect. So I think it's better to return here.